File: Luogu/P_1387_最大正方形.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, m;
    cin >> n >> m;
    vector<vector<int>> M(n, vector<int>(m));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> M[i][j];
        }
    }
    vector<vector<int>> f(n + 1, vector<int>(m + 1, 0));
    int ans = 0;
    for (int i = n - 1; i >= 0; i--) {
        for (int j = m - 1; j >= 0; j--) {
            if (M[i][j] == 0) continue;
            f[i][j] = min({f[i + 1][j], f[i][j + 1], f[i + 1][j + 1]}) + 1;
            ans = max(ans, f[i][j]);
        }
    }
    cout << ans << endl;
    return 0;
}

File: Luogu/P_1387_最大正方形.py
from functools import cache

n, m = map(int, input().split())
M = [list(map(int, input().split())) for _ in range(n)]

@cache
def dfs(i, j):
    if i >= n or j >= m:
        return 0
    if M[i][j] != 1:
        return 0
    return 1 + min(dfs(i + 1, j), dfs(i, j + 1), dfs(i + 1, j + 1))

ans = 0
for i in range(n):
    for j in range(m):
        ans = max(ans, dfs(i, j))
print(ans)